packages/zk-certificates/contracts/HumanIDSaltRegistry.sol

```diff
@@ -0,0 +1,179 @@
+// SPDX-License-Identifier: BUSL-1.1
+pragma solidity ^0.8.7;
+pragma abicoder v2;
+
+import {GuardianRegistry} from './GuardianRegistry.sol';
+
+struct SaltLockingZkCert {
+    uint256 zkCertId;
+    address guardian;
+    uint256 expirationTime;
+    bool revoked;
+}
+
+struct UserData {
+    // this is the registered salt hash for the user
+    uint256 saltHash;
+    // this is the list of zkCerts that might block resetting the salt
+    uint256[] zkCerts;
+}
+
+/**
+ * @title HumanIDSaltRegistry
+ * @author Galactica dev team
+ * @notice Registry to limit a user to one salt to achieve a unique and single dApp specific humanID.
+ */
+contract HumanIDSaltRegistry {
+    GuardianRegistry public guardianRegistry;
+    address public zkCertRegistry;
+
+    // mapping from  id hash to the salt hash
+    mapping(uint256 => UserData) internal _userData;
+    mapping(uint256 => SaltLockingZkCert) internal _saltLockingZkCerts;
+
+    constructor(address guardianRegistry_, address zkCertRegistry_) {
+        guardianRegistry = GuardianRegistry(guardianRegistry_);
+        zkCertRegistry = zkCertRegistry_;
+    }
+
+    /**
+     * @notice Register a salt hash for a user. Reverts if another salt is registered already. Saves zkCert data for eventual recovery.
+     * @param zkCert - Data for the SaltLockingZkCert.
+     * @param idHash - Hash identifying the user. It is supposed to be the poseidon Hash of the name, birthday and citizenship.
+     * @param saltHash - Hash of the salt.
+     */
+    function onZkCertIssuance(
+        SaltLockingZkCert calldata zkCert,
+        uint256 idHash,
+        uint256 saltHash
+    ) external {
+        require(
+            msg.sender == zkCertRegistry,
+            'HumanIDSaltRegistry: only zkCertRegistry can call this function'
+        );
+
+        _registerSaltHash(idHash, saltHash);
+
+        _userData[idHash].zkCerts.push(zkCert.zkCertId);
+        _saltLockingZkCerts[zkCert.zkCertId] = zkCert;
+    }
+
+    /**
+     * @notice Inform salt registry about zkCert revocation. This helps resetting the salt before zkCert expiration if needed.
+     * @param zkCertId - Id of the zkCert to be revoked.
+     */
+    function onZkCertRevocation(uint256 zkCertId) external {
```

Comment:
can we already remove the revoked ZkCert from the ZkCerts array in Userdata to make the loop in getSaltLockingZkCert more efficient?

Reply:
I think this is not worth the effort here because `getSaltLockingZkCert` is only needed on-chain in the unlikely case that a user can not recover his account.